### PR TITLE
mediatek: filogic: add Keenetic KN-3911 support

### PR DESCRIPTION
**Pull request by Linaro1985**
Specification:
 - MT7981 CPU using 2.4GHz and 5GHz WiFi (both AX)
 - 512MB RAM
 - 128MB SPI NAND
 - 2 LEDs (green, orange)
 - 3 buttons (fn, reset, wps)
 - 2 2.5Gbit ethernet ports based on Airoha EN8811H phy

Serial Interface:
 - 3 Pins GND, RX, TX
 - Settings: 115200, 8N1

Notes:
 - The device supports dual boot mode

Flash instruction:
The only way to flash OpenWrt image is to use tftp recovery mode in U-Boot:

1. Configure PC with static IP 192.168.1.2/24 and tftp server.
2. Rename "openwrt-mediatek-filogic-keenetic_kn-3911-squashfs-factory.bin"
   to "KN-3911_recovery.bin" and place it in tftp server directory.
3. Connect PC with ethernet port, press the reset button, power up
   the device and keep button pressed until status led start blinking.
4. Device will download file from server, write it to flash and reboot.

### CODE REVIEW

File: target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+		phy-handle = <&phy11>;
+		phy-mode = "2500base-x";
+		nvmem-cell-names = "mac-address";
+		nvmem-cells = <&macaddr_factory_4 1>;
+	};
+};
+
+&mdio_bus {
+	phy10: ethernet-phy@10 {
+		compatible = "ethernet-phy-id03a2.a411";
```
[the remark this answers is not in the document]
Comment (dangowrt) [reply target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts]: Please try if everything works also with `compatible = "ethernet-phy-ieee802.3-c22";`

[the remark this answers is not in the document]
Comment (Linaro1985) [reply target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts]: @dangowrt works well in both cases.
...
Thu Oct 31 15:14:43 2024 kern.info kernel: [    5.195171] Airoha EN8811H mdio-bus:0a: MD32 firmware version: 24082302
Thu Oct 31 15:14:43 2024 kern.info kernel: [    7.392059] Airoha EN8811H mdio-bus:0b: MD32 firmware version: 24082302
...
Thu Oct 31 15:14:46 2024 kern.info kernel: [   18.719051] mtk_soc_eth 15100000.ethernet lan: PHY [mdio-bus:0a] driver [Airoha EN8811H] (irq=53)
Thu Oct 31 15:14:46 2024 kern.info kernel: [   18.733917] mtk_soc_eth 15100000.ethernet lan: configuring for phy/2500base-x link mode
Thu Oct 31 15:14:46 2024 kern.info kernel: [   18.808647] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:0b] driver [Airoha EN8811H] (irq=38)
Thu Oct 31 15:14:46 2024 kern.info kernel: [   18.817997] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
...

File: target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+	phy11: ethernet-phy@b {
+		reg = <0xb>;
+
+		interrupt-parent = <&pio>;
+		interrupts = <23 IRQ_TYPE_EDGE_FALLING>;
+		reset-gpios = <&pio 27 GPIO_ACTIVE_LOW>;
+		reset-assert-us = <10000>;
+		reset-deassert-us = <20000>;
+
+		phy-mode = "2500base-x";
```
Comment (aiamadeus) [inline target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts:142]: Same here.
Besides, I think it looks better this way:
```
ethernet-phy@x {
		reg = <x>;
		reset-assert-us = xxx
		reset-deassert-us = xxx
		reset-gpios = xxx
		interrupt-parent = xxx
		interrupts = xxx
};
```

File: target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+	phy10: ethernet-phy@a {
+		reg = <0xa>;
+
+		interrupt-parent = <&pio>;
+		interrupts = <38 IRQ_TYPE_EDGE_FALLING>;
+		reset-gpios = <&pio 22 GPIO_ACTIVE_LOW>;
+		reset-assert-us = <10000>;
+		reset-deassert-us = <20000>;
+
+		phy-mode = "2500base-x";
```
[the remark this answers is not in the document]
Comment (Linaro1985) [reply target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts:128]: @aiamadeus what about this https://github.com/openwrt/openwrt/blob/main/target/linux/mediatek/dts/mt7981b-openwrt-one.dts#L166 ?

[the remark this answers is not in the document]
Comment (Linaro1985) [reply target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts]: > ```
> phy-mode = "2500base-x";
> full-duplex;
> pause;
> ```
> 
> These are useless and can be removed.

Yes, it works without this lines. Thanks!

File: target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+				label = "config_2";
+				reg = <0x7600000 0x80000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&xhci {
+	status = "disabled";
```
Comment (aiamadeus) [inline target/linux/mediatek/dts/mt7981b-keenetic-kn-3911.dts:285]: It is disabled by default.

File: target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac [hunk 5]
```diff
@@ -96,6 +96,10 @@ case "$board" in
 		[ "$PHYNBR" = "0" ] && echo "$addr" > /sys${DEVPATH}/macaddress
 		[ "$PHYNBR" = "1" ] && macaddr_add $addr 1 > /sys${DEVPATH}/macaddress
 		;;
+	keenetic,kn-3911)
```
Comment (aiamadeus) [inline target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:99]: Please follow the alphabetical order.

File: target/linux/mediatek/filogic/base-files/etc/board.d/02_network [hunk 6]
```diff
@@ -94,6 +94,7 @@ mediatek_setup_interfaces()
 	mercusys,mr90x-v1)
 		ucidef_set_interfaces_lan_wan "lan0 lan1 lan2" eth1
 		;;
+	keenetic,kn-3911|\
```
Comment (aiamadeus) [inline target/linux/mediatek/filogic/base-files/etc/board.d/02_network:97]: Same here.

**Comment by Linaro1985** [comment]
@dangowrt Hi. Maybe I forgot to do something? If so, please tell me. What else do I need to do to get this PR accepted?

**Comment by hauke** [comment]
It looks like this board supports two flash banks. Do you have to make sure that it boots up from the first one?

**Comment by Crav88** [comment]
Thanks for your work :)  
image

**Comment by Crav88** [comment]
> Thanks for your work :) image

@Linaro1985 do you check the wifi performance? Apparently the speed it's just stuck on 1G.

**Comment by Linaro1985** [comment]
Hi @Crav88
No, I haven't tested it to see if it's possible to get speeds greater than 1G.

**Comment by Crav88** [comment]
With the Keenetic os yes it’s possibile , now 160mhz … like this https://www.speedtest.net/result/i/6553703714

apparently bad

**Comment by Crav88** [comment]
if you use it like Ap all good... @Linaro1985

**Pull request merged**